File: app/src/main/java/com/dhbikoff/breakout/PhaseBase.java
package com.dhbikoff.breakout;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;

import java.util.ArrayList;

public interface PhaseBase {
    PhaseBase prepare();
    void update(float eventX);
    void draw(Canvas canvas);
}

/** PhaseManager **/
class PhaseManager {
    static OpeningPhase openingPhase;
    static BuildPhase buildPhase;
    static PlayPhase playPhase;
    static ClosingPhase closingPhase;
    static ClaarPhase clearPhase;
    static PhaseBase currentPhase = new NonePhase();

    public static void init(Context context, boolean soundflg, int screenW, int screenH) {
        openingPhase = new OpeningPhase();
        buildPhase = new BuildPhase(context, soundflg, screenW, screenH);
        playPhase = new PlayPhase();
        closingPhase = new ClosingPhase();
        clearPhase = new ClaarPhase();
        currentPhase = openingPhase;
    }

    public static PhaseBase getPhase() {
        currentPhase = currentPhase.prepare();
        return currentPhase;
    }
}

/** NonePhase(何もしないフェーズ) **/
class NonePhase implements PhaseBase {
    @Override public PhaseBase prepare() { return this; }
    @Override public void update(float eventX) { }
    @Override public void draw(Canvas canvas) { }
}
/** OpeningStage **/
class OpeningPhase implements PhaseBase {
    private final Paint paint = new Paint();
    private static int phaseCounter = 0;
    /* コンストラクタ */
    public OpeningPhase() {
        paint.setTextAlign(Paint.Align.CENTER);
        paint.setColor(Color.WHITE);
        paint.setTextSize(72);
    }
    /* 準備 */
    @Override public PhaseBase prepare() {
        final int OPENING_PHASE_PERIOD = 66;
        phaseCounter++;
        if(phaseCounter < OPENING_PHASE_PERIOD)
            return this;
        else
            phaseCounter = 0;
        return PhaseManager.buildPhase;
    }
    /* 更新 */
    @Override public void update(float eventX) {/*何もしない*/}
    /* 描画 */
    @Override
    public void draw(Canvas canvas) {
        canvas.drawText("GET READY...",(float)canvas.getWidth()/2, (float)(canvas.getHeight()/2) , paint);
    }
}

/** BuildStage **/
class BuildPhase implements PhaseBase {
    private Context mContext;
    private final boolean mSoundFlg;
    private final int mScreenW;
    private final int mScreenH;
    /* コンストラクタ */
    public BuildPhase(Context context, boolean soundFlg, int screenW, int screenH) {
        mContext = context;
        mSoundFlg= soundFlg;
        mScreenW = screenW;
        mScreenH = screenH;
    }
    /* 準備 */
    @Override
    public PhaseBase prepare() {
        Ball ball = new Ball(mContext, mSoundFlg);
        ball.init(mScreenW, mScreenH);
        Paddle paddle = new Paddle();
        paddle.init(mScreenW, mScreenH);
        ArrayList<Block> blocksList = new ArrayList<>();
        initBlocks(mScreenW, mScreenH, blocksList);
        PhaseManager.playPhase.setParams(ball, paddle, blocksList);
        mContext = null;
        return PhaseManager.playPhase;
    }
    @Override public void update(float eventX) { }
    @Override public void draw(Canvas canvas) {}
    /* 準備2(ブロック初期化) */
    private void initBlocks(int screenW, int screenH, ArrayList<Block> blocksList) {
        int blockHeight=  screenW / 36;
        int spacing    =  screenW / 144;
        int topOffset  =  screenH / 10;
        int blockWidth = (screenW / 10) - spacing;

        for (int idxy = 0; idxy < 10; idxy++) {
            for (int idxx = 0; idxx < 10; idxx++) {
                int y_coordinate = (idxy * (blockHeight + spacing)) + topOffset;
                int x_coordinate = idxx * (blockWidth + spacing);

                Rect rect = new Rect();
                rect.set(x_coordinate, y_coordinate, x_coordinate + blockWidth, y_coordinate + blockHeight);

                if(idxy < 2)      blocksList.add(new Block(rect, Color.RED));
                else if(idxy < 4) blocksList.add(new Block(rect, Color.YELLOW));
                else if(idxy < 6) blocksList.add(new Block(rect, Color.GREEN));
                else if(idxy < 8) blocksList.add(new Block(rect, Color.MAGENTA));
                else              blocksList.add(new Block(rect, Color.LTGRAY));
            }
        }
    }
}

/** PlayStage **/
class PlayPhase implements PhaseBase {
    private Ball ball = null;
    private Paddle paddle = null;
    private ArrayList<Block> blocksList = null;
    private int score = 0;
    private int lifeGauge = 5;
    public final Paint scorePaint;
    public final Paint lifePaint;
    private boolean gameoverFlg = false;

    public PlayPhase() {
        scorePaint = new Paint();
        scorePaint.setColor(Color.WHITE);
        scorePaint.setTextSize(60);
        lifePaint = new Paint();
        lifePaint.setTextAlign(Paint.Align.RIGHT);
        lifePaint.setColor(Color.WHITE);
        lifePaint.setTextSize(60);
    }

    public void setParams(Ball ball, Paddle paddle, ArrayList<Block> blocksList) {
        this.ball = ball;
        this.paddle = paddle;
        this.blocksList = blocksList;
    }

    @Override
    public PhaseBase prepare() {
        if(blocksList.isEmpty()) {
            gameoverFlg = false;
            return PhaseManager.clearPhase;
        }
        else if(!gameoverFlg)
            return this;
        else {
            gameoverFlg = false;
            return PhaseManager.closingPhase;
        }
    }

    @Override
    public void update(float eventX) {
        /* ボール更新 */
        lifeGauge -= ball.updateCoordinates();
        if(lifeGauge <= 0)
            gameoverFlg = true;
        /* パドル移動 */
        paddle.move((int)eventX);
        /* コリジョンチェック(ボール-パドル) */
        ball.checkPaddleCollision(paddle);
        /* コリジョンチェック(ボール-ブロックs) */
        score += ball.checkBlocksCollision(blocksList);
    }

    @Override
    public void draw(Canvas canvas) {
        /* ボール描画 */
        ball.draw(canvas);
        /* ブロック描画 */
        for(Block item : blocksList)
            item.draw(canvas);
        /* パドル描画 */
        paddle.draw(canvas);

        canvas.drawText("SCORE = "+ score, 0, 50, scorePaint);
        canvas.drawText("Life: " + lifeGauge, canvas.getWidth(), 50, lifePaint);
    }
}

/** ClaarStage **/
class ClaarPhase implements PhaseBase {
    final Paint paint = new Paint();
    public ClaarPhase() {
        paint.setTextAlign(Paint.Align.CENTER);
        paint.setTextSize(72);
        paint.setColor(Color.GREEN);
    }

    @Override public PhaseBase prepare() { return this; }
    @Override public void update(float eventX) {/*何もしない*/}
    @Override
    public void draw(Canvas canvas) {
        canvas.drawText("Game Cleared!", (float)canvas.getWidth()/2,(float)(canvas.getHeight()/2)-50, paint);
        canvas.drawText("Good Gob!!!", (float)canvas.getWidth()/2,(float)(canvas.getHeight()/2)-50+72, paint);
    }
}

/** ClosingStage **/
class ClosingPhase implements PhaseBase {
    final Paint paint = new Paint();
    public ClosingPhase() {
        paint.setTextAlign(Paint.Align.CENTER);
        paint.setTextSize(72);
        paint.setColor(Color.RED);
    }

    @Override public PhaseBase prepare() { return this; }
    @Override public void update(float eventX) {/*何もしない*/}
    @Override
    public void draw(Canvas canvas) {
        canvas.drawText("GAME OVER!!!", (float)canvas.getWidth()/2,(float)(canvas.getHeight()/2)-50, paint);
    }
}

File: app/src/main/java/com/aaa/breakout/PhaseBase.kt
package com.aaa.breakout

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect

const val OPENING_PHASE_PERIOD = 66

interface PhaseBase {
    fun prepare(context: Context): PhaseBase
    fun update(eventX: Float)
    fun draw(canvas: Canvas)
}

/** PhaseManager  */
internal object PhaseManager {
    private lateinit var openingPhase: OpeningPhase
    lateinit var buildPhase: BuildPhase
    lateinit var playPhase: PlayPhase
    lateinit var closingPhase: ClosingPhase
    lateinit var clearPhase: ClearPhase
    private var currentPhase: PhaseBase = NonePhase()

    fun init(soundFlg: Boolean, screenW: Int, screenH: Int) {
        openingPhase= OpeningPhase()
        buildPhase  = BuildPhase(soundFlg, screenW, screenH)
        playPhase   = PlayPhase()
        closingPhase= ClosingPhase()
        clearPhase  = ClearPhase()
        currentPhase= openingPhase
    }

    fun getNextPhase(context: Context): PhaseBase {
        currentPhase = currentPhase.prepare(context)
        return currentPhase
    }

    /** NonePhase(何もしないフェーズ)  */
    class NonePhase : PhaseBase {
        override fun prepare(context: Context): PhaseBase { return this }
        override fun update(eventX: Float) {}
        override fun draw(canvas: Canvas) {}
    }

    /** OpeningStage  */
    class OpeningPhase : PhaseBase {
        private val paint = Paint()

        /* 初期化 */
        init {
            paint.textAlign = Paint.Align.CENTER
            paint.color = Color.WHITE
            paint.textSize = 72f
        }

        /* 準備 */
        override fun prepare(context: Context): PhaseBase {
            phaseCounter++
            if (phaseCounter < OPENING_PHASE_PERIOD) return this
            else phaseCounter = 0
            return buildPhase
        }

        /* 更新 */
        override fun update(eventX: Float) { /*何もしない*/}

        /* 描画 */
        override fun draw(canvas: Canvas)
            = canvas.drawText("GET READY...", canvas.width.toFloat() / 2,(canvas.height / 2).toFloat(), paint)

        companion object {
            private var phaseCounter = 0
        }
    }

    /** BuildStage  */
    class BuildPhase /* コンストラクタ */(
        private val mSoundFlg: Boolean,
        private val mScreenW: Int,
        private val mScreenH: Int
    ) : PhaseBase {
        /* 準備 */
        override fun prepare(context: Context): PhaseBase {
            val ball   = Ball(context, mSoundFlg).apply { init(mScreenW, mScreenH) }
            val paddle = Paddle().apply { init(mScreenW, mScreenH) }
            val blocksList: ArrayList<Block> = ArrayList()
            initBlocks(mScreenW, mScreenH, blocksList)
            playPhase.setParams(ball, paddle, blocksList)
            return playPhase
        }

        override fun update(eventX: Float) {}
        override fun draw(canvas: Canvas) {}

        /* 準備2(ブロック初期化) */
        private fun initBlocks(screenW: Int, screenH: Int, blocksList: ArrayList<Block>) {
            val blockHeight = screenW / 36
            val spacing = screenW / 144
            val topOffset = screenH / 10
            val blockWidth = (screenW / 10) - spacing

            for (idxy in 0..9) {
                for (idxx in 0..9) {
                    val ycoordinate = (idxy * (blockHeight + spacing)) + topOffset
                    val xcoordinate = idxx * (blockWidth + spacing)

                    val rect = Rect()
                    rect.set(xcoordinate, ycoordinate, xcoordinate + blockWidth, ycoordinate + blockHeight)

                    if (idxy < 2) blocksList.add(Block(rect, Color.RED))
                    else if (idxy < 4) blocksList.add(Block(rect, Color.YELLOW))
                    else if (idxy < 6) blocksList.add(Block(rect, Color.GREEN))
                    else if (idxy < 8) blocksList.add(Block(rect, Color.MAGENTA))
                    else blocksList.add(Block(rect, Color.LTGRAY))
                }
            }
        }
    }

    /** PlayStage  */
    internal class PlayPhase : PhaseBase {
        private lateinit var ball: Ball
        private lateinit var paddle: Paddle
        private lateinit var blocksList: ArrayList<Block>
        private var score = 0
        private var lifeGauge = 5
        private var scorePaint: Paint = Paint().apply {
                                                    color = Color.WHITE
                                                    textSize = 60f
                                                }
        private val lifePaint: Paint = Paint().apply {
                                                    textAlign = Paint.Align.RIGHT
                                                    color = Color.WHITE
                                                    textSize = 60f
                                                }
        private var gameoverFlg = false

        fun setParams(ball: Ball, paddle: Paddle, blocksList: ArrayList<Block>) {
            this.ball = ball
            this.paddle = paddle
            this.blocksList = blocksList
        }

        override fun prepare(context: Context): PhaseBase {
            if (blocksList.isEmpty()) {
                gameoverFlg = false
                return clearPhase
            }
            else if (!gameoverFlg)
                return this
            else {
                gameoverFlg = false
                return closingPhase
            }
        }

        override fun update(eventX: Float) {
            /* ボール更新 */
            lifeGauge -= ball.updateCoordinates()
            if (lifeGauge <= 0) gameoverFlg = true
            /* パドル移動 */
            paddle.move(eventX.toInt())
            /* コリジョンチェック(ボール-パドル) */
            ball.checkPaddleCollision(paddle)
            /* コリジョンチェック(ボール-ブロックs) */
            score += ball.checkBlocksCollision(blocksList)
        }

        override fun draw(canvas: Canvas) {
            /* ボール描画 */
            ball.draw(canvas)
            /* ブロック描画 */
            for (item in blocksList) item.draw(canvas)
            /* パドル描画 */
            paddle.draw(canvas)

            canvas.drawText("SCORE = $score", 0f, 50f, scorePaint)
            canvas.drawText("Life: $lifeGauge", canvas.width.toFloat(), 50f, lifePaint)
        }
    }

    /** ClearStage  */
    internal class ClearPhase : PhaseBase {
        private val paint: Paint = Paint().apply {
                                        textAlign = Paint.Align.CENTER
                                        textSize = 72f
                                        color = Color.GREEN
                                    }
        override fun prepare(context: Context): PhaseBase { return this }
        override fun update(eventX: Float) { /*何もしない*/ }
        override fun draw(canvas: Canvas) {
            canvas.drawText( "Game Cleared!", canvas.width.toFloat() / 2, (canvas.height / 2).toFloat() - 50, paint )
            canvas.drawText( "Good Gob!!!", canvas.width.toFloat() / 2, (canvas.height / 2).toFloat() - 50 + 72, paint )
        }
    }

    /** ClosingStage  */
    internal class ClosingPhase : PhaseBase {
        private val paint: Paint = Paint().apply {
                                            textAlign = Paint.Align.CENTER
                                            textSize = 72f
                                            color = Color.RED
                                        }

        override fun prepare(context: Context): PhaseBase { return this }
        override fun update(eventX: Float) { /*何もしない*/ }
        override fun draw(canvas: Canvas)
            = canvas.drawText("GAME OVER!!!", canvas.width.toFloat() / 2, (canvas.height / 2).toFloat() - 50, paint)
    }
}
